add typescript types

Zemnmez opened the pull request:
adds some typescript types for this package. would like better examples to test from – I don't use this package directly, just one of its far deps and I'm trying to work my way up adding types

Zemnmez committed:
add types

Zemnmez committed:
add comment for Marker

Zemnmez committed:
add testing for type checks

ChristianMurphy commented on package.json:
```diff
@@ -44,7 +47,8 @@
     "test-coverage": "nyc --reporter lcov tape test.js",
-    "test": "npm run format && npm run build && npm run test-coverage"
+    "test": "npm run format && npm run build && npm run test-coverage && npm run typecheck",
+    "typecheck": "tsc"
```
Could we do additional type checking with dtslint?
for example https://github.com/Rokt33r/hast-util-to-html/blob/318112f2c59275c59d8b687f75f4c21b12486f44/types/tslint.json
https://github.com/syntax-tree/hast-util-to-html/blob/0fef5aa542e1d8a9361f3a0ea8e38adf07fa870f/package.json#L60-L61

Zemnmez replied:
yes!

ChristianMurphy commented on package.json:
```diff
@@ -15,12 +15,15 @@
   ],
   "dependencies": {
+    "typescript": "^3.7.4",
+    "unist-util-is": "^4.0.1",
```
Could/should these be devDependencies?

Zemnmez replied:
they should. Tried to add them with `--dev` (like `yarn`) guess it doesnt work the same...

ChristianMurphy replied:
yarn uses `--dev`, npm uses `--save-dev`

ChristianMurphy commented on tsconfig.json:
```diff
@@ -0,0 +1,66 @@
+{
```
Could we use this config: https://github.com/Rokt33r/hast-util-to-html/blob/318112f2c59275c59d8b687f75f4c21b12486f44/types/tsconfig.json ?

Zemnmez replied:
will just need to add `"noemit": true` because i have type tests that dont compile to js

ChristianMurphy replied:
Using dtslint to do the checking should remove the need for `"noemit": true`

Zemnmez replied:
that's weird but ok haha

ChristianMurphy commented on index.d.ts:
```diff
@@ -0,0 +1,93 @@
+import { Test } from 'unist-util-is';
```
Could we include typescript files in the prettier formatter in package.json?
So that it matches the code style.

ChristianMurphy replied:
Also could we include the typescript version required as a comment?
similar to https://github.com/syntax-tree/hast-util-to-html/blob/0fef5aa542e1d8a9361f3a0ea8e38adf07fa870f/types/index.d.ts#L1

ChristianMurphy reviewed (commented):
Thanks @Zemnmez !
https://github.com/syntax-tree/hast-util-to-html/pull/17 is an example of the structure that has been used in other unified repos.
In particular the TSConfig and DTSLint config.

Zemnmez committed:
Improve types, lint types.

- test now imports itself as a package (neat!)
- dtslint checks the types are pretty
- `typescript` and `unist-util-is` moved to devdeps
(should have been before)
- maybe this is known, but weirdly this package depends on an older
version of itself through remark-preset-wooorm

Zemnmez commented:
@ChristianMurphy looks like the tslint you're using doesn't reflect travis' – it seems to have "semicolon" and "whitespace" turned on

Zemnmez committed:
make lint rules pass
- remember to add tslint.json

Zemnmez commented:
what's more ... I don't think it's going to be possible to make this work without either (1) changing travis to use this eslint config or (2) dropping prettier – as the prettier config is actually stripping the semicolons that travis is complaining about

ChristianMurphy commented:
@Zemnmez we've been disabling those rules on the tslint side to let prettier handle the formatting.
https://github.com/Rokt33r/hast-util-to-html/blob/318112f2c59275c59d8b687f75f4c21b12486f44/types/tslint.json

Rokt33r commented on index.d.ts:
```diff
@@ -0,0 +1,104 @@
+import {Test} from 'unist-util-is'
+import * as unist from 'unist'
+
+export {}
```
Empty export statement. Please remove this line

Zemnmez replied:
the empty export statement is there so that explicit exporting is turned on. otherwise the Null type would be exported

Rokt33r replied:
This is CJS module. Only namespace and the function will be exported. 

How `Null` can be exported?

Rokt33r replied:
<img width="885" alt="Screen Shot 2020-01-19 at 0 17 00" src="https://user-images.githubusercontent.com/5865853/72675389-0e021180-3a51-11ea-8194-2570291124de.png">

As you can see in the screenshot, `Null` is not accessible unless it is defined in the namespace.

Zemnmez replied:
Odd... this was something dtslint asked for. I'll remove it and we'll see what happens.

Rokt33r commented on index.d.ts:
```diff
@@ -0,0 +1,104 @@
+type Null = undefined | null
+
+declare namespace messageControl {
+  export {}
```
This one is also useless

Zemnmez replied:
same as above

Rokt33r commented on index.d.ts:
```diff
@@ -0,0 +1,104 @@
+declare namespace messageControl {
+  export {}
+  /** a comment marker */
+  export interface Marker<N extends unist.Node> {
```
`export`  in namespace is useless. Please delete it.

Zemnmez replied:
same as above

Rokt33r commented on index.d.ts:
```diff
@@ -0,0 +1,104 @@
+    (node: N): Marker<N> | Null
+  }
+
+  interface MessageControlOptionsWithReset<T extends unist.Node>
```
I don't think we need separated interfaces based on `reset` option.
 `enable` and `disable` are optional regardless they are used or not.

Zemnmez replied:
I don't really understand. the reset option is documented to change the interpretation of other fields, which this expresses. Doing it this way means a discriminated union can intuit the state of the other fields. see: https://www.typescriptlang.org/docs/handbook/advanced-types.html#discriminated-unions

Rokt33r replied:
The spec documentation doesn't said `disabled` or `enabled` must not be provided if `reset` is `true`, `false` or `undefined`.

Zemnmez replied:
Catching up on my types here ... I have `enable?: undefined` (should maybe be `undefined | null`), but that just specifies that if enable *is* specified it should be undefined (i.e. not that the field should not be present, only that if it is present it should be `undefined`), which I think makes sense for ensuring well-formed `MessageControlOptions`. This doesn't mean that `MessageControlOptions` can't have `enable` as some other value, but in this case you'd need to do a type guard to extract that.

I'd really much rather communicate this relationship through types than rely on documentation, so maybe setting these to `unknown` when they have no defined meaning in the spec is a good halfway house?

Rokt33r commented on index.d.ts:
```diff
@@ -0,0 +1,104 @@
+
+  /** parse a possible comment marker node to a Marker */
+  export interface MarkerFn<N extends unist.Node> {
+    (node: N): Marker<N> | Null
```
`MarkerFn` should return `null` not `null` or `undefined`.

Zemnmez replied:
Happy to change this but it seems to intersect with the discussion on how `null` is / should be treated in the unified ecosystem

Rokt33r commented on index.d.ts:
```diff
@@ -0,0 +1,104 @@
+
+declare namespace messageControl {
+  export {}
+  /** a comment marker */
```
Try to keep using same description from readme.

```ts
/**
 * A comment marker.
 */
interface Marker<N extends unist.Node>{
```

Zemnmez replied:
Is the issue here the missing punctuation?

Rokt33r commented on index.d.ts:
```diff
@@ -0,0 +1,104 @@
+    reset?: false | Null
+
+    enable?: string[] | Null
+    disable?: Null
```
@ChristianMurphy Do you think we need to care about `null`?
AFAIK, we've never allow `null` for option prop in type definitions because we're using `Partial`. 

```ts
type Partial<T> = {
    [P in keyof T]?: T[P];
};
```

ChristianMurphy replied:
I don't think we need `null` as part of the type.
Also with `Partial` the `?` annotation is not needed.
In addition in the docs for this plugin:
> options.disable
> List of ruleIds to turn on if reset: false (Array.<string>, optional).

I think `disable` should have a `string[]` included in the typing

Zemnmez replied:
Partial makes all the fields optional, but the documentation states that only some of the fields are optional. Null is used here for the same reason as https://github.com/DefinitelyTyped/DefinitelyTyped/pull/41631 -- the use of `null` vs `undefined` seems to be interchangeable within the unist ecosystem. In standard typescript null and undefined are interchangeable, but it seems that now it's considered best practice to separate them. see: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#--strictnullchecks

Zemnmez replied:
(in this case, the DefinitelyTyped type definitions were actually *rejecting* the mdast examples because they didn't allow `null`)

Rokt33r replied:
AFAIK, handling `null` as `undefined` is a handy casual convention but not a spec in unified ecosystem. So, unless our specs said explicitly that `null` is allowed, I would go for using `?` only without caring `null`.

Rokt33r replied:
@wooorm How do you think? Should we explicitly allow `null` type for optional props or ignore `null` type like how we've done before?

wooorm replied:
I think in programming it is wise to stay close to the language: JavaScript is loose, so when in JS, embrace that and be loose as well. E.g., if expecting a string and a number is passed in, that number should be serialised to string. Same with `null` and `undefined` (I believe this is called “nully”) when something is optional.
Now, there are two competing viewpoints here: a) TS is strict; b) TS should be close to the implementation.
I don’t know which one should have more weight! 🤷‍♂️

ChristianMurphy replied:
I'd lean towards more strict.

If we go in the direction of supporting everything javascript does, we may also need to add all the falsy values to the typing: `false`, `0`, `NaN`, `0n`, and `''`.
While that technically may be supported, I'm not sure that's something we want to encourage adopters to leverage.

Zemnmez replied:
The issue here is while I appreciate the strictness, I feel strongly that I don't want to type anything that would break existing code in the unified ecosystem, as noted on my previous comment. The DefinitelyTyped standard is to say if the type could be potentially `null` or `undefined`, and I think where it was never strictly specified canonically previously I'd like to call out that `null` and `undefined` should be expected.

If we don't do this, then anyone without the relevant TypeScript flag is going to get errors when `null` is encountered instead of `undefined` or vice-versa ... the TypeScript system will make sure you do (!field) as a guard instead of `field !== undefined` if we specify both `null` and `undefined`.

On `false`, `0`, `NaN` etc, I think there's a point where it gets unreasonable – I don't think many would say that these can / should / are used interchangeably with undefined in the unified ecosystem and as such I don't think it makes sense to support it.

ChristianMurphy replied:
> I feel strongly that I don't want to type anything that would break existing code in the unified ecosystem

Adding typings is considered a breaking change in the Unified ecosystem.
The main reason, is that the typings we provide target TypeScript 3+, and some adopters use TypeScript v1 and v2.

Given that it is already a potentially breaking change, we include any other potentially breaking changes that may improve clarity at the same time.

> The DefinitelyTyped standard is to say if the type could be potentially `null` or `undefined`

Do you have a citation for this?
This is not included in the best practice guide https://definitelytyped.org/guides/best-practices.html or the contributing guide https://definitelytyped.org/guides/contributing.html

> If we don't do this, then anyone without the relevant TypeScript flag is going to get errors when null is encountered instead of undefined or vice-versa ... the TypeScript system will make sure you do `(!field)` as a guard instead of `field !== undefined` if we specify both `null` and `undefined`.

Yes, that is what a type checker does.

Zemnmez replied:
@ChristianMurphy it's very clearly stated here

https://github.com/DefinitelyTyped/DefinitelyTyped#some-packages-have-no-tslintjson-and-some-tsconfigjson-are-missing-noimplicitany-true-noimplicitthis-true-or-strictnullchecks-true 

> Yes, that is what a type checker does.
i ... don't really understand your response. I know what a typechecker does. My point is that if we **do not** specify `null`, typescript may allow `(field !== undefined)`. Since this does not exclude `null`, runtime bugs can and will happen because `null` fields will be considered 'defined'.

ChristianMurphy replied:
> https://github.com/DefinitelyTyped/DefinitelyTyped#some-packages-have-no-tslintjson-and-some-tsconfigjson-are-missing-noimplicitany-true-noimplicitthis-true-or-strictnullchecks-true

Not it does not.
This states that three compiler options should be enabled.
1. These compiler options are already included in the unified typings, as part of `"strict": true`
2. It states this should be done for typing themselves, it does not apply to downstream adopters. Meaning if adopters want a looser interpretation of `null` and `undefined` the unified API the can have it.
3. This speaks only to the compiler options, not to whether project should explicitly type with `null`, `undefined` or both.

> i ... don't really understand your response. I know what a typechecker does.

It means you stated what a type checker will do in a specific circumstance, _by itself_, this information has no bearing on the conversation.
It would have bearing if:

- choosing just undefined would break `unified-message-control` itself
- choosing just `undefined` would mislead adopters
- choosing just `undefined` would change the stated API

With my current knowledge, I don't believe any of those three is true in this case.
This code supports `undefined`, stating so would not mislead adopters, and `null` is only explicitly supported as part of the API for the `unist-util-is` test.

> My point is that if we do not specify null, typescript may allow `(field !== undefined)`. Since this does not exclude null, runtime bugs can and will happen because null fields will be considered 'defined'.

Yes, but where would that happen?
All the potentially nullable values are going into `unified-message-controller`, not coming out.
What happens inside `unified-message-controller` is an implementation detail, only callbacks and return values would have to potential to impact adopters.

Rokt33r commented on index.d.ts:
```diff
@@ -0,0 +1,104 @@
+
+export {}
+
+type Null = undefined | null
```
I don't think we need this.
Rather than using Null, we can just do like the below example.

```ts
someStrKey?: string | null
```

Rokt33r replied:
But don't fix this now. I think we need to discuss https://github.com/unifiedjs/unified-message-control/pull/7#discussion_r368238256  first.

Rokt33r commented on index.d.ts:
```diff
@@ -0,0 +1,104 @@
+// on any Node which is clearly *not* what we want
+
+declare function messageControl<T extends unist.Node>(
+  // tslint:disable-next-line:no-unnecessary-generics
```
Please apply this rule in tslint.json like

https://github.com/wooorm/parse-entities/pull/17/files#diff-65d6de72f3f0a20f873033fe5b908bb6R6

Rokt33r reviewed (changes requested):
Please use same description from the original documentation. If you want to edit something, I think we should fix the original one first.

Zemnmez commented:
> @Zemnmez we've been disabling those rules on the tslint side to let prettier handle the formatting.
> https://github.com/Rokt33r/hast-util-to-html/blob/318112f2c59275c59d8b687f75f4c21b12486f44/types/tslint.json

These rules are already disabled, but it looks (looked?) like travis was rejecting because of this lint rule regardless

https://github.com/unifiedjs/unified-message-control/pull/7/files#diff-ace19bd0c04529e685320269e3c05de9

Zemnmez committed:
Change 'Null' to 'Nullish'

Zemnmez committed:
Move generics lint rule to tslint.json

Zemnmez commented:
also not in my commit messages: fields which lose their meaning conditional on others are now `unknown` instead of `null | undefined`

Zemnmez commented:
Removing the `export {}` broke the travis build; going to disable `strict-export-declare-modifiers` to align with the discussion before.

```
ERROR: 6:6    strict-export-declare-modifiers  All declarations in this module are exported automatically. Prefer to explicitly write 'export' for clarity. If you have a good reason not to export this declaration, add 'export {}' to the module to shut off automatic exporting. See: https://github.com/Microsoft/dtslint/blob/master/docs/strict-export-declare-modifiers.md
ERROR: 8:19   strict-export-declare-modifiers  All declarations in this module are exported automatically. Prefer to explicitly write 'export' for clarity. If you have a good reason not to export this declaration, add 'export {}' to the module to shut off automatic exporting. See: https://github.com/Microsoft/dtslint/blob/master/docs/strict-export-declare-modifiers.md
ERROR: 96:18  strict-export-declare-modifiers  All declarations in this module are exported automatically. Prefer to explicitly write 'export' for clarity. If you have a good reason not to export this declaration, add 'export {}' to the module to shut off automatic exporting. See: https://github.com/Microsoft/dtslint/blob/master/docs/strict-export-declare-modifiers.md
ERROR: 101:1  strict-export-declare-modifiers  All declarations in this module are exported automatically. Prefer to explicitly write 'export' for clarity. If you have a good reason not to export this declaration, add 'export {}' to the module to shut off automatic exporting. See: https://github.com/Microsoft/dtslint/blob/master/docs/strict-export-declare-modifiers.md
```

Zemnmez committed:
Turn off strict-export-declare-modifiers (pull #7)

ChristianMurphy commented:
> also not in my commit messages: fields which lose their meaning conditional on others are now `unknown` instead of `null | undefined`

What is the thinking behind this?
Conditional types can be represented by creating a union of the different variations, or by making the properties optional.
How does `unknown` come into the mix?

Zemnmez commented:
> > also not in my commit messages: fields which lose their meaning conditional on others are now `unknown` instead of `null | undefined`
> 
> What is the thinking behind this?
> Conditional types can be represented by creating a union of the different variations, or by making the properties optional.
> How does `unknown` come into the mix?

https://github.com/unifiedjs/unified-message-control/pull/7#discussion_r370272934

Previously, these fields which have no specified value when other fields are not present were set to `null | undefined`, but as noted these fields are neither specified to have `null | undefined` value when they are not specified to be present, so I'm using `unknown`. If the user wants to extract the data from these fields, they can narrow it.

ChristianMurphy commented on index.d.ts:
```diff
@@ -0,0 +1,101 @@
+
+// this tslint rule makes no sense to me...
+// if T was any we'd essentially have *zero* typechecking
+// on any Node which is clearly *not* what we want
```
comments explaining the code are welcome.
commentary on lint rules in the code doesn't offer much benefit to maintainers or adopters.

ChristianMurphy commented on tsconfig.json:
```diff
@@ -0,0 +1,15 @@
+    "esModuleInterop": true,
+
+    // this looks super weird, but baseUrl doesn't consider .
+    // a module so we have to explicitly tell it in paths...
```
It's not too weird, many definitely typed repos have this configuration.
and most unified ecosystem repos do.

Zemnmez committed:
remove vestigial comment about generics lint rule

ChristianMurphy commented on index.d.ts:
```diff
@@ -0,0 +1,96 @@
+
+  /** parse a possible comment marker node to a Marker */
+  interface MarkerFn<N extends unist.Node> {
+    (node: N): Marker<N> | Nullish
```
I don't think this can be `null` or `undefined`.
It will only be called if there is a value.

ChristianMurphy committed:
typings: refactor to support unified

also:
* removes nullish
* adds comments
* adds more type tests
* clarifies which props can be used with reset

ChristianMurphy commented:
Did some local testing, came across an issue where with the current typings, `unified-message-control` cannot be used as a unified plugin.

```js
var vfile = require('to-vfile')
var report = require('vfile-reporter')
var remark = require('remark')
var control = require('unified-message-control')
var mdastMarker = require('mdast-comment-marker')

remark()
  .use(warn)
  .use(control, {name: 'foo', marker: mdastMarker, test: 'html'})
  .process(vfile.readSync('example.md'), function(err, file) {
    console.error(report(err || file))
  })

function warn() {
  return function(tree, file) {
    file.message('Whoops!', tree.children[1], 'foo:thing')
  }
}
```

I opened https://github.com/Zemnmez/unified-message-control/pull/1 to resolve this and to include a couple other minor changes.

Zemnmez committed:
Merge pull request #1 from ChristianMurphy/types/refactor

typings: refactor to support unified

Zemnmez committed:
fix stray backquote

Zemnmez committed:
capitalise field comments

ChristianMurphy reviewed (approved):
Looks good, thanks @Zemnmez! :bow:

Zemnmez commented:
https://github.com/unifiedjs/unified-message-control/pull/7#discussion_r370640811
`"strictNullChecks": true` being specified as *having* to be explicitly specified in all DefinitelyTyped definitions reads to me as a pretty strong statement of preference for explicitly specifying `null` vs `undefined`. We have functions in this repo that will return `null`, not `undefined`.

There are two outcomes here:
1) We set `strictNullChecks: true` and specify where nulls are possible.
2) We set `strictNullChecks: false` and allow nulls and undefined to be equivalent.

The other outcome, where `strictNullChecks: true`, and nulls are unspecified is very worrying for me, as I've personally seen typings with this configuration break canonical examples in the unified ecosystem (see my MDAST typing PR).

I'm done fighting on this, though. Whatever you want goes.

Zemnmez commented:
@ChristianMurphy's fork of the typings are super slick btw. very clean

ChristianMurphy commented:
> strong statement of preference for explicitly specifying `null` vs `undefined`.

I agree,
Definitely typed does prefer strict separation between `null` and `undefined`.
That doesn't answer the question of does this API accept both.
Some example from other APIs in unified and syntax-tree
https://github.com/syntax-tree/unist-util-is/blob/8c0a649bcb202738584dbd81b84b19cc25e3561c/index.d.ts#L61-L62
https://github.com/syntax-tree/mdast-util-toc/blob/a6361867d908bdac1db70baaff566b3fee733a96/types/index.d.ts#L8-L56
Or any from https://github.com/search?l=TypeScript&q=org%3Asyntax-tree+TypeScript+Version&type=Code

`?` is used to make fields optional, but `null` was not explicitly added.

> We have functions in this repo that will return `null`, not `undefined`.

Yes, and that is reflected in the typing.

> I've personally seen typings with this configuration break canonical examples in the unified ecosystem (see my MDAST typing PR)

I agree there may be places in unist/mdast/hast where `null` should be explicitly included in the type.
I'd be careful conflating the syntax trees with plugin settings though.

The content in syntax trees are often user generated, go through many layers of plugins and transforms, and it is common for adopters to build custom plugins to handle specific use cases.
There are parts of the syntax tree APIs which do explicitly support `null`, and this should be reflected to ensure smooth adoption for typescript developers.
Being more open with the typing, while still doing our best to guide people in the right direction, has value for syntax trees.

Plugin settings are generally not have user generated content feeding into them, they aren't fed forward to future plugins, they are used once, and while they may use `!` over `!== undefined` in the code for brevity, they generally do not explicitly document `null` as part of the API.
Being more strict with the typing has value here.
